src/backend/distributed/utils/resource_lock.c

```diff
@@ -119,3 +121,58 @@ UnlockJobResource(uint64 jobId, LOCKMODE lockmode)
 
 	LockRelease(&tag, lockmode, sessionLock);
 }
+
+
+/*
+ * LockShards takes shared locks on the metadata and the data of all shards in
+ * shardIntervalList. This prevents concurrent placement changes and concurrent
+ * DML statements that require an exclusive lock.
+ */
+void
+LockShards(List *shardIntervalList, LOCKMODE lockMode)
+{
+	ListCell *shardIntervalCell = NULL;
+
+	/* lock shards in order of shard id to prevent deadlock */
+	shardIntervalList = SortList(shardIntervalList, CompareShardIntervalsById);
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		int64 shardId = shardInterval->shardId;
+
+		/* prevent concurrent changes to number of placements */
+		LockShardDistributionMetadata(shardId, lockMode);
+
+		/* prevent concurrent update/delete statements */
+		LockShardResource(shardId, lockMode);
+	}
+}
+
+
+/*
+ * CompareShardIntervalsById is a comparison function for sort shard
+ * intervals by their shard ID.
+ */
+int
+CompareShardIntervalsById(const void *leftElement, const void *rightElement)
```

Comment:
I recently discovered we have a utils/shardinterval_utils.c file now, which seems like the appropriate place for this function.